### DISCUSSION TITLE
It should be possible to create pulses from an array

### DISCUSSION THREAD

**Discussion by fabiques** [qilimanjaro-tech/qililab#352] [Ideas]
In order to have full flexibility and control of our systems we need the possibility to send arbitrarily shaped pulses down to the chip.

If we, as physicists, think of a new pulse shape that in someway could improve our experiments, we should have the freedom to implement it.

The current implementation, although more practical for most cases, is too restrictive.

**Comment by fabiques** [comment]
@rsagas this relates to the problems you @paulqili and @andrea-pema were having today

**Comment by fedonman** [comment]
I totally agree with you. The design of the `PulseOperation` class, that will be used in place of current `Pulse` class in the new circuit, was done with that in mind.

`PulseOperation` is the base class that defines some common methods, e.g. `envelope()`, and then each sub-class is responsible for implementing it. There is a family of sub-classes that create this envelope from a mathematical description, e.g. Gaussian, Square, CosineSquared, Sinc etc. However, there will be simply another sub-class, probably called `ArbitraryPulse` that will allow to create any pulse directly from a numpy array.
